shelley/chain-and-ledger/executable-spec/src/Shelley/Spec/Ledger/Value.hs

```diff
@@ -0,0 +1,407 @@
+gt :: Value crypto -> Value crypto -> Bool
+-- If both are zero then checkBinRel will be vacuously true. So we have a special case.
+gt = checkBinRel (>)
+
+{-# INLINABLE leq #-}
+-- | Check whether one 'Value' is less than or equal to another. See 'Value' for an explanation of how operations on 'Value's work.
+leq :: Value crypto -> Value crypto -> Bool
+-- If both are zero then checkBinRel will be vacuously true, but this is fine.
+leq = checkBinRel (<=)
+
+{-# INLINABLE lt #-}
+-- | Check whether one 'Value' is strictly less than another. See 'Value' for an explanation of how operations on 'Value's work.
+lt :: Value crypto -> Value crypto -> Bool
+-- If both are zero then checkBinRel will be vacuously true. So we have a special case.
+lt = checkBinRel (<)
+
+{-# INLINABLE eq #-}
+-- | Check whether one 'Value' is equal to another. See 'Value' for an explanation of how operations on 'Value's work.
+eq :: Value crypto -> Value crypto -> Bool
+-- If both are zero then checkBinRel will be vacuously true, but this is fine.
+eq = checkBinRel (==)
+
+-- | Split a value into its positive and negative parts. The first element of
+--   the tuple contains the negative parts of the value, the second element
+--   contains the positive parts.
+--
+--   @negate (fst (split a)) `plus` (snd (split a)) == a@
+-- TODO
+-- split :: Value crypto -> (Value crypto, Value crypto)
+-- split (Value mp) = (negate (Value neg), Value pos) where
+--   (neg, pos) = Map.mapThese splitIntl mp
+--
+--     splitIntl :: Map.Map TokenName Integer -> These (Map.Map TokenName Integer) (Map.Map TokenName Integer)
+--     splitIntl mp' = These l r where
+--       (l, r) = Map.mapThese (\i -> if i <= 0 then This i else That i) mp'
+
+-- TODO do this right
+splitValueFee :: Value crypto -> Integer -> (Value crypto, Coin)
+splitValueFee v _ = (v, Coin 0)
+-- (Value v) n =
+--   | n <= 0 = error "must split coins into positive parts"
+--   | otherwise = (fmap $ n `div` m, Coin $ n `rem` m)
+
+-- CBOR
+
+instance
+  (Crypto crypto)
+  => ToCBOR (CompactValue crypto)
+ where
+   toCBOR = \case
+     AdaOnly c ->
+           encodeListLen 2
+           <> toCBOR (0 :: Word8)
+           <> toCBOR c
+     MixValue (Value v) ->
+           encodeListLen 2
+           <> toCBOR (1 :: Word8)
+           <> toCBOR v
+
+instance
+  (Crypto crypto)
+  => FromCBOR (CompactValue crypto)
+ where
+  fromCBOR = do
+    decodeWord >>= \case
+      0 -> do
+        c <- fromCBOR
+        pure $ AdaOnly c
+      1 -> do
+        v <- fromCBOR
+        pure $ MixValue $ Value v
+      k -> invalidKey k
+
+
+instance
+  (Crypto crypto)
+  => ToCBOR (Value crypto)
+ where
+   toCBOR = (\case
+     AdaOnly c ->
+           encodeListLen 2
+           <> toCBOR (0 :: Word8)
+           <> toCBOR c
+     MixValue (Value v) ->
+           encodeListLen 2
+           <> toCBOR (1 :: Word8)
+           <> toCBOR v) . valueToCompactValue
```

Comment:
This still makes no sense! You've got the instance for `Value` defined in terms of `CompactValue` and vice versa!

Reply:
Ah, I'd missed that this was unpacking the Value constructor here.